### PR TITLE
Add zookeeper retries for exists check in forcefullyRemoveBrokenOutdatedPartFromZooKeeper

### LINKED ISSUE #59426: loadOutdatedDataParts: Terminate called for uncaught exception:
Report: https://s3.amazonaws.com/clickhouse-test-reports/59003/600eae00ccb0f5328ea66cad2045049bf7b67a91/stress_test__msan_.html

Trace:
```
 [ 29426 ] {} <Fatal> BaseDaemon: (version 24.1.1.1, build id: B443CA84A08DB930CF665314DC54DDD8016C2B24, git hash: 78990185ccb58b29593b8e7b28b64a0cc5cbe01b) (from thread 29668) Terminate called for uncaught exception:
 [ 29426 ] {} <Fatal> BaseDaemon: Code: 999. Coordination::Exception: Session expired. (KEEPER_EXCEPTION), Stack trace (when copying this message, always include the lines below):
 [ 29426 ] {} <Fatal> BaseDaemon: 
 [ 29426 ] {} <Fatal> BaseDaemon: 0. ./contrib/llvm-project/libcxx/include/exception:134: Poco::Exception::Exception(String const&, int) @ 0x00000000409ee289 in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 1. ./build_docker/./src/Common/Exception.cpp:96: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x00000000200d2513 in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 2. ./src/Common/Exception.h:0: Coordination::Exception::Exception<char const (&) [16]>(char const (&) [16], Coordination::Error) @ 0x000000003ddf5091 in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 3. ./src/Common/ZooKeeper/IKeeper.h:0: Coordination::ZooKeeper::pushRequest(Coordination::ZooKeeper::RequestInfo&&) @ 0x000000003de709d0 in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 4. ./build_docker/./src/Common/ZooKeeper/ZooKeeperImpl.cpp:1325: Coordination::ZooKeeper::exists(String const&, std::function<void (Coordination::ExistsResponse const&)>, std::shared_ptr<std::function<void (Coordination::WatchResponse const&)>>) @ 0x000000003de769df
 [ 29426 ] {} <Fatal> BaseDaemon: 5. ./contrib/llvm-project/libcxx/include/__functional/function.h:818: ? @ 0x000000003dd61f81 in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 6. ./contrib/llvm-project/libcxx/include/__functional/function.h:818: ? @ 0x000000003dd6113a in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 7. ./build_docker/./src/Common/ZooKeeper/ZooKeeper.cpp:513: zkutil::ZooKeeper::existsWatch(String const&, Coordination::Stat*, std::function<void (Coordination::WatchResponse const&)>) @ 0x000000003dd62490 in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 8. ./build_docker/./src/Common/ZooKeeper/ZooKeeper.cpp:497: zkutil::ZooKeeper::exists(String const&, Coordination::Stat*, std::shared_ptr<Poco::Event> const&) @ 0x000000003dd5a9d1 in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 9. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:7644: DB::StorageReplicatedMergeTree::forcefullyRemoveBrokenOutdatedPartFromZooKeeperBeforeDetaching(String const&) @ 0x000000003a2368c5 in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 10. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:815: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::MergeTreeData::loadOutdatedDataParts(bool)::$_0, void ()>>(std::__function::__policy_storage c
 [ 29426 ] {} <Fatal> BaseDaemon: 11. ./contrib/llvm-project/libcxx/include/__functional/function.h:0: ? @ 0x0000000033e03f6a in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 12. ./contrib/llvm-project/libcxx/include/future:1694: std::__packaged_task_func<std::function<std::future<void> (std::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, String
 [ 29426 ] {} <Fatal> BaseDaemon: 13. ./contrib/llvm-project/libcxx/include/future:2069: std::packaged_task<void ()>::operator()() @ 0x000000003188107d in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 14. ./contrib/llvm-project/libcxx/include/__functional/function.h:717: ? @ 0x0000000033e045ae in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 15. ./base/base/../base/wide_integer_impl.h:809: ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x00000000204043cf in /usr/bin/clickhouse
 [ 29426 ] {} <Fatal> BaseDaemon: 16. ./build_docker/./src/Common/ThreadPool.cpp:0: ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::\'lambda0\'()>(v
 [ 29426 ] {} <Fatal> BaseDaemon: 17. ./contrib/llvm-project/libcxx/include/__functional/function
```

We need to capture exceptions. And ideally have retries.

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Closes https://github.com/ClickHouse/ClickHouse/issues/59426.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 208a91a2f7ae97795e03c62c1cce121775ee58ff with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61324/208a91a2f7ae97795e03c62c1cce121775ee58ff/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -7680,9 +7680,13 @@ void StorageReplicatedMergeTree::forcefullyRemoveBrokenOutdatedPartFromZooKeeper
     /// It could became Outdated because it was merged/mutated (so we have a covering part) or because it was dropped.
     /// But DROP [PART]ITION waits for all Outdated parts to be loaded, so it's not the case.
 
-    auto zookeeper = getZooKeeper();
+    bool exists = false;
     String part_path = replica_path + "/parts/" + part_name;
-    if (!zookeeper->exists(part_path))
+    ZooKeeperRetriesInfo retries_info{/* max_retries */5, /* initial_backoff_ms */100, /* max_backoff_ms */100};
```
Comment (kssenii) [inline src/Storages/StorageReplicatedMergeTree.cpp:7685]: Not sure if a new setting needs to be introduced for these, currently there are settings for keeper retries in backups and in inserts only.

Comment (Algunenano) [reply src/Storages/StorageReplicatedMergeTree.cpp:7685]: I would use the insert settings, we did the same for system.zookeeper. We should probably unify both insert and backup settings and have a universal set of settings that works for any query.

**Comment by antonio2368** [comment]
But this still doesn't explain the reason why we ended up with uncaught exception right?
This will just reduce chance of it happening.

**Comment by kssenii** [comment]
> But this still doesn't explain the reason why we ended up with uncaught exception right?
This will just reduce chance of it happening.

yes, but see my comment https://github.com/ClickHouse/ClickHouse/issues/59426#issuecomment-1995078559

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**